### ISSUE TITLE
RokPad Functionality for Custom HTML Particle

### ISSUE THREAD

**Issue by DamirPecnik** [gantry/gantry5#236]
Some users have been asking for RokPad Functionality in Custom HTML Particle.

**Comment by w00fz** [comment]
This is kind of a duplicate of GANTRYFIVE-66 and GANTRYFIVE-67, good to have the issue here on GitHub tho.
We are going to improve the field to use a more user friendly editor as well as extend its functionality to support Markdown and HTML (possibly WYSIWYG)

**Comment by rhukster** [comment]
We definitely plan on having some editor options in future Gantry releases.  I personally want a markdown editors :)

**Comment by mahagr** [comment]
I already implemented simple editor to test how Joomla editors would work in G5. It is working -- now and then, but unfortunately because of our AJAX calls and overrides, it is not as simple task as I expected it to be.

**Comment by w00fz** [comment]
I think that rather than trying to make it work with the platforms editors that is going to be very hard to support (been there, done that), we should just focus on providing our owns: 
1. Simple Textarea
2. RokPad like editor with markdown, html, css, twig support and a live view for html/markdown
3. WYSIWYG like editor (this is going to be the most complicated one and in my opinion we shouldn't do it but rather rely on the Preview of the RokPad like editor)

**Comment by DamirPecnik** [comment]
My suggestion is this:
2. RokPad like editor with markdown, html, css, twig support and a live view for html/markdown

**Comment by newkind** [comment]
2 would be awesome, but only if it would be available for all platforms and not Joomla only ;)

**Comment by mahagr** [comment]
We would already have working WYSIWYG editor in Joomla if we didn't use AJAX calls everywhere.

**Comment by w00fz** [comment]
My point is we shouldn't even look at implementing platforms Editor but just provide our own.
It makes things much easier to manage, the way platforms implement their editors is just too different between them and will be a nightmare.
We had these issues a lot, even for the simplest things like RokBox injecting into the editor a prebuilt string, it had to be pretty much rewritten for Wordpress because it is completely different than in Joomla.

If we provide our own editors, it will be automatically universal AND work even for those platforms where there is no fancy editors.

**Comment by DamirPecnik** [comment]
i agree with @w00fz

**Comment by newkind** [comment]
agreed!

**Comment by mahagr** [comment]
I can agree on that!

**Comment by rhukster** [comment]
+1

**Comment by BjoernH01** [comment]
+1

**Comment by marktaylor46** [comment]
+1 members are really needing a good wysiwyg editor for customhtml particles (and other particles where you can key in html). They need it because most of there clients are frightened of code and need to maintain their own content. The lack of a wysiwyg editor is perceived as a inhibitor to the take up of Gantry 5. Here is an example of what I mean: 
http://www.rockettheme.com/forum/gantry-framework/243951-gantry-5-yikes?start=10#1229021

**Comment by mahagr** [comment]
Already implemented in 5.1.2, closing.

**Comment by topwebs** [comment]
Clients will not input or edit raw html. The number of options does get confusing. Do I use a Joomla custom html module that can utilize an editor? Or a G5 module with a custom html particle? Or a G5 custom html particle, neither of the last two can currently utilize an editor? I think you will need to make WYSIWYG editors work with Joomla at least.

**Comment by hennysmafter** [comment]
We definitely need a good WYSIWYG editor, and markup and raw code. It would be even best if users could switch between those without having to reopen a given particle. This issue is now almost open for a year. And I agree many more important stuff has been added to Gantry but even on the Rockettheme forums the people are complaining because the end users don't have editors. So devs please make this a first priority as this is the nr: 1 complain our company receives from our customers.

**Comment by alisamii** [comment]
I absolutely agree. I think an interface similar to how the editor has been implemented on the RocketTheme forums is ideal...with possible the options to have wysiwyg, github-style markdown and raw options.

**Comment by alisamii** [comment]
A really nice, clean, elegant editor that could be added to Gantry is the Froala WYSIWYG Editor

https://www.froala.com/wysiwyg-editor

It might be a good option.

**Comment by blizam** [comment]
+1

**Comment by w00fz** [comment]
That wysiwyg editor you linked is not free, requires a license.

**Comment by Mineslu** [comment]
I think the correct path is to create a particle named "Custom html" and to show a block loading the wysiwyg editor that was defined as preset in the global Joomla configurations (JCE or other).
That is my opinion.

**Comment by marktaylor46** [comment]
Well from a Joomla perspective I'd agree but really we need a solution that works consistently across all the CMS platforms that Gantry 5 works with so that would push you down the route of an editor provided by Gantry 5 admin.

**Comment by Mineslu** [comment]
OK.
Here there some:
https://www.tinymce.com/

**Comment by alisamii** [comment]
@w00fz true, Froala Editor isn't free.

Some nice alternatives are
http://quilljs.com
http://getcontenttools.com
http://summernote.org

**Comment by paulhibbitts** [comment]
Might https://simplemde.com/ be another candidate too? It is already being used in the Grav 'Add Page by Form' plugin too. https://github.com/bleutzinn/grav-plugin-add-page-by-form

**Comment by tjcreadev** [comment]
[TinyMCE for Gantry.zip](https://github.com/gantry/gantry5/files/1008465/TinyMCE.for.Gantry.zip)

Sharing this upon MrT's recommendation to me on the RT forum:

I was able to get TinyMCE to work within Gantry 5's layout manager on Joomla. I have it on 6 of our websites now and have done various component and cms updates - still working.

I created a custom base.html.twig file in the custom/admin/templates/partial folder for my template with a few changes from the default base.html.twig file. Within the custom file I included locally hosted TinyMCE .js files (they need to be hosted locally for them to work correctly) and then initiated the TinyMCE editor on textarea focus (due to how the particle fields load on the page when editing a particular particle - using 'focus' was the best solution I could get to work). I also used TinyMCE's setup function to populate the actual textarea with the content placed within the iframe box that TinyMCE creates - otherwise, your textarea would remain empty and nothing would save properly.

Attached is a zip file I created with my solution. It can be extracted in the custom folder of a Gantry 5 Template (at least on Joomla).

I'm planning on continuing to work on it more myself to clean it up a bit, but it's at least a start for what I was wanting it to do.

**Comment by Mineslu** [comment]
Great, I'll try it and tell you my experience.

**Comment by Mineslu** [comment]
![fireshot capture 34 - your joomla site - ad_ - http___localhost_masterpets2_administrator_index php](https://cloud.githubusercontent.com/assets/9434043/26166857/a1899b98-3afa-11e7-92d9-e7a9d46147e5.png)

Great,
One question, what changes have you made in the TinyMCE?
Thank

**Comment by tjcreadev** [comment]
In the editor itself I have tried all the buttons (plugins) with exception to the image button.

I didn't make any adjustments to the TinyMCE files. They are in the same state as when downloaded from TinyMCE website.

**Comment by tjcreadev** [comment]
One thing to note:
TinyMCE seems to have some issue with Grammarly plugin. If a word is misspelled or has some other alert from Grammarly the TinyMCE formatting will not work.

**Comment by tjcreadev** [comment]
Some updates to what I've done. I will try to get another zip file uploaded soon.

1 - I have created a second TinyMCE initiation for inputs (for things like titles where you aren't using a full fledge textarea). I edited the particle input field and assigned it a class for TinyMCE to initiate based on. For my own uses I have created some custom particle fields that I can assign to inputs that I do not wish to have additional formatting options (alt text, class, etc).

2 - I've been able to attach a file manager plugin to TinyMCE that links to a Gantry 5 templates 'custom/images' folder. The file manager allows for some great image editing (crop, resize, color adjustments, etc).

For my own purposes, this allows me to set my clients up with a Gantry 5 template where they will not - for the most part - need to use Joomla articles and page content particles. This lets them stay in a single administrative area in Joomla which is great for the less tech-saavy clients.

**Comment by Mineslu** [comment]
Great job

**Comment by Mineslu** [comment]
I have a big problem with the publisher.
For example, I want the "Source code" button to allow me to hide the editor and manually modify the code. But when I insert the editor and test it I get an error. Has it happened to you? I attach some images of the problem.
![1](https://cloud.githubusercontent.com/assets/9434043/26271706/1240a054-3ccf-11e7-95c6-ad535e04ceff.jpg)
![2](https://cloud.githubusercontent.com/assets/9434043/26271707/12463884-3ccf-11e7-840f-c3a83f4c55b7.jpg)

**Comment by tjcreadev** [comment]
I ran into the same issue at first. I have fixed it and it will be included in my updated .zip file

**Comment by Mineslu** [comment]
Excellent. If it's not too much trouble, when will you publish the update?
Thank you

**Comment by tjcreadev** [comment]
I don't be back at my computer until tomorrow morning (currently 9:40pm my time). I should have a new .zip file uploaded tomorrow.

**Comment by Mineslu** [comment]
Thank!!

**Comment by tjcreadev** [comment]
Here are two .zip files containing my updated solution for TinyMCE within the Gantry 5 layout manager. I still have a long way to go before I'm happy with the final result, but it's a big step forward from my initial solution.

**INSTRUCTIONS:**

1 - Upload and extract the filemanager .zip to your site's root folder.

2 - Upload and extract the Gantry5TinyMCE .zip to your Gantry5's template folder.

3 - For this to work in it's current state you need to make two minor adjustments to the config.php file located within the <site root>/filemanager/config/ folder:

On lines 76 and 85 you will need to adjust the 'upload_dir' and 'current_path' options to match your own template's name. So in this case replace '/templates/g5_ember/custom/images/' with your own template name ('/templates/<YOUR-TEMPLATE-NAME>/custom/images/') in the directory path. Otherwise you'll get an error when using the file manager in the TinyMCE toolbar.

----------------

I hope to find a solution that intuitively detects the template name and assigns the correct path in the future, but unless I'm missing something obvious... it will be a difficult thing to tackle. Alternatively, the image upload folder could just be linked to Joomla's image folder so it can be easily extracted and the .config file wouldn't need touched. This would allow Gantry 5's image picker to access the images you upload since it pulls in Joomla's images, but it wouldn't allow you to access images uploaded via Gantry's image picker from the TinyMCE toolbar. 

There is a minor bug with the file manager I'm going to have to look into further. When you first load the file manager the files are available, but the thumbnails don't appear. If you click the 'refresh' icon or if you do any other action within the file manager they will appear. It's just the initial load that seems to have an issue.

[filemanager-EXTRACT_IN_PUBLIC_HTML.zip](https://github.com/gantry/gantry5/files/1016306/filemanager-EXTRACT_IN_PUBLIC_HTML.zip)
[Gantry5TinyMCE-May20-2017-EXTRACT_IN_GANTRY5_TEMPLATE_FOLDER.zip](https://github.com/gantry/gantry5/files/1016307/Gantry5TinyMCE-May20-2017-EXTRACT_IN_GANTRY5_TEMPLATE_FOLDER.zip)

**Comment by mahagr** [comment]
@tjcreadev If you go to `System - Gantry 5` plugin settings, there's an option to use Joomla images folder instead of theme.

**Comment by tjcreadev** [comment]
@manage

Sometimes the simplest approach is right in front of you and the right way to go. That will probably help me a good bit on this.

Thanks!

**Comment by Mineslu** [comment]
Great job, I'm using it right now in my projects.

**Comment by Mineslu** [comment]
Hello,
I'm working with the plugin that you have provided us and everything is going great.
But I found an inconvenience.
When I go to the Modules manager and I try to create one of the type "Gantry particle", Then I choose for example, the particle: "CUSTOM HTML" is no longer loaded Tinymce, as you can see in the image.
Thank
![fireshot capture 39 - modulo_ gantry 5 parti_ - http___localhost_masterpets2_administrator_index php](https://user-images.githubusercontent.com/9434043/26857516-a7a009f8-4af0-11e7-9833-3f8b67f13fce.png)

**Comment by tjcreadev** [comment]
Yes, I see the problem. I am calling the TinyMCE function from a file within the Gantry 5 dashboard. I will see about calling it elsewhere to allow it to work here as well.

Thanks for the find!

**Comment by Mineslu** [comment]
Thank!

**Comment by mahagr** [comment]
The issue comes from the AJAX request, eg. the page doesn't have TinyMCE loaded and AJAX response cannot add new scripts. Could be fixed by preloading the JS files from the particle field.

**Comment by Mineslu** [comment]
Hi,
Do I think that gantry already supports fields with editors?
As I see it here:
administrator/components/com_gantry5/templates/forms/fields/textarea/editor.html.twig

But when I try to create my field in my particle, like this:
    html:
      type: textarea.editor
      label: Contenido
      description: Enter custom HTML into here.

I get this:

![your joomla site - administracion](https://user-images.githubusercontent.com/9434043/27087119-bc0e8706-5019-11e7-8813-5c1b8d8d5522.jpg)

**Comment by tjcreadev** [comment]
@Mineslu As of now, Gantry doesn't already support it. Maybe that was a leftover from the devs trying it (I haven't looked).

I have been swamped with client projects and have been unable to devote time to this in a few weeks. I will try to invest some time in it soon.

**Comment by Mineslu** [comment]
Thank  my friends. @tjcreadev

**Comment by mahagr** [comment]
Yeah, I was playing out with editors a long time ago and just left the field into there. It likely doesn't work anymore.

**Comment by Mineslu** [comment]
Thank my friends.

**Comment by openelearning** [comment]
Hello,
Anyone know if there is solutions to have HTML Editor as particle field for wordpress ?
Thanks

**Comment by tjcreadev** [comment]
> Hello,
> Anyone know if there is solutions to have HTML Editor as particle field for wordpress ?
> Thanks

I spent a couple hours around 18 months ago trying to get my solution to work on WordPress, but something was preventing it. I didn't pursue it after that. I would like to get it to work for WP though.

I'm currently using my Tinymce solution on around 250 websites and it's been working great for my needs. I've added a few more features to my latest version - some are TinyMCE options and some are custom developed:

- Can pull in accent colors from styles tab of Gantry 5 for text color and even a few custom button formats
- Additional options (several button block types, image/file manager, bootstrap layout manager within text editor)

It is still just a pet project though - I'd be excited to see what the Gantry team comes up with for permanent use.

**Comment by thexmanxyz** [comment]
I'm also think it is finally time to implement this feature. It would be a big leap if we could at least have support for an editor within textfields.

**Comment by viperovm** [comment]
@tjcreadev 
Hi! You've done a great job with WYSIWYG editor. Thank you very much!
I've uploaded all the folders and made all changes and it doesn't work unfortunately. I'm sure that I'm doing something wrong so I hope you will help me with this issue.
This is part of my yaml file:

`.introtext:
          type: textarea.tinyMCEtextarea
          label: Introtext`

Any information you need I'll provide you.
Many thanks in advance!

**Comment by thexmanxyz** [comment]
@viperovm If you added `.introtext: type: textarea.tinyMCEtextarea label: Introtext` in one line it wouldn't work at all. In YAML files you need to use correct indention (2-4 spaces I think for sub elements). If you already did that I'm out :D. I just assume because you pasted it this way here.

**Comment by viperovm** [comment]
<img width="457" alt="Снимок экрана 2019-07-24 в 12 15 20" src="https://user-images.githubusercontent.com/53254495/61785782-b752ac80-ae0c-11e9-9733-bc1e81b44a47.png">
It has the same look as yours. But still thanks)

**Comment by thexmanxyz** [comment]
@viperovm Alright! That looks fine so it must be something different.

**Comment by tjcreadev** [comment]
Hi,

It's been a while since I've worked with the versions I've posted here. I've changed several things since then, but this may solve your issue:

Instead of "textarea.tinyMCEtextarea" as the type - try "textarea.textareaEditor"

As an FYI - this should work for particles within Gantry 5's dashboard and layout manager, but it doesn't work outside of it.

Thanks!

**Comment by mediaDESIGN-SK** [comment]
@tjcreadev:
Thank you for the effort and sharing of your solution to the editor problem in Gantry5.
This works quite well in the Gantry Layout Manager. Unfortunately it doesn't work for particles that are managed in the Joomla module manager. Is there a solution?

**Comment by tjcreadev** [comment]
> @tjcreadev:
> Thank you for the effort and sharing of your solution to the editor problem in Gantry5.
> This works quite well in the Gantry Layout Manager. Unfortunately it doesn't work for particles that are managed in the Joomla module manager. Is there a solution?

I haven't spent time working on that particular item. Honestly, I was hoping and kind of expecting an official solution to this to have come out by now. If I end up with some time soon for some R&D on this project I will see what I can figure out.

Thanks!

**Comment by pau1a** [comment]
Hi guys,
Is this close to being solved at all?
I mostly use joomla with rokpad and like to create complex html modules all over the place.
I thought Id try Gantry for Wordpress and am I right in thinking that the only way to do this is to create my own new particles and text edit them on the backend at the twig file?
I feel like Im missing something. 
Im happy enough doing things this way as I just name the particles per the use case I need them for but it dont half seem a bit of a kludge.

**Comment by blizam** [comment]
@pau1a what exactly is your issue? You can certainly maintain custom HTML inside a TWIG file and perhaps add backend fields to the particle settings as needed... I've definitely written multiple custom particles that are effectively static, with no backend options, with everything contained withing the TWIG file.

Alternatively you can place custom html, TWIG code, and WP shortcodes directly in the standard Custom Particle that comes with Gantry.

**Comment by pau1a** [comment]
Hey Matt, Thanks for getting back to me. Theres no specific issue. I wondered if the process of creating custom html blocks in the WP site had been given a fully functioning editor yet. If Im using hydrogen for example on Joomla, I end up basically writing each block as I go and I mostly find it easiest to do that via rokpad as a custom HTML module. With WP it seems that that the facility to edit fullscreen etc isnt there so Ive basically, as you say been writing individual particles for each block I want to create and drawing the SCSS from the custom system. Its no bother to work this way as Im as comfortable with vi or nano as anything else but like I say, I wondered if there had been progress.

**Comment by blizam** [comment]
Gotcha, yeah, this thread here with the TinyMCE solution for Joomla is the only thing I'm presently aware of. The author of that notes above that he tried to get it to work with WordPress but wasn't fully successful with it. I myself will use the built in Custom particle while developing something but always just copy+pasting from a local text editor. Sometimes I end up leaving it that way and other times I think it justifies being it's own custom Particle and I'll write the yaml file for it and move it in the the /particles folder. The ability to drop WP shortcodes in to the Custom particle (or via a twig file too) can be very powerful as well as so many other WP things use shortcodes. I had one site where the client wanted to use WPBakery and they would build elements and drop the generated shortcodes (which basically all just translated into pre-built HTML tags) into a custom particle.

**Comment by tjcreadev** [comment]
I will try to address the WP issue with my TinyMCE solution in the coming months. I've primarily switched over to JAMStack solutions for client projects, but Gantry is still a great solution when I'm working with any Joomla/WP projects.

**Comment by andreascarfo** [comment]
Decades after... Just to remember something we all need.

**Comment by tjcreadev** [comment]
Wow. I have been so far departed from Gantry development these past few years that I forgot about this entirely. I can't believe there is still no official rich text editor integration. I don't plan on expanding on what I did (nearly a decade ago - yeesh), but someone else is welcome to take it and run. Heck, there is probably a better way to do it - i don't even remember what I coded for it back in 2017 😅
